### PR TITLE
[Sparse] GCNII example

### PR DESCRIPTION
**Pull request by paoxiaode**
## Description
The GCNII example code https://github.com/dmlc/dgl/issues/4932 using sparse matrix API
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
Ref to the full-supervised.py in the repo https://github.com/chennnM/GCNII
### Performace
|dataset |  val acc| test acc|
| - |   -|  -|
| Cora |  0.8 | 0.85 |

### CODE REVIEW

File: examples/sparse/gcnii.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    # Training settings
+    argparse = argparse.ArgumentParser()
+    argparse.add_argument("--epochs", type=int, default=1500)
+    argparse.add_argument("--lr", type=float, default=0.01)
+    argparse.add_argument("--wd", type=float, default=5e-4)
+    argparse.add_argument("--n_layers", type=int, default=64)
+    argparse.add_argument("--hidden_size", type=int, default=64)
+    argparse.add_argument("--dropout", type=float, default=0.5)
+    argparse.add_argument("--alpha", type=float, default=0.2)
+    argparse.add_argument("--lamda", type=float, default=0.5)
```
Comment (mufeili) [inline examples/sparse/gcnii.py:138]: Simplify the implementations to set default values whenever needed and remove `argparse`.

File: examples/sparse/gcnii.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    test_acc = (pred[test_mask] == label[test_mask]).float().mean()
+    return val_acc, test_acc
+
+
+def train(args, model, g, A_norm, H):
+    label = g.ndata["label"]
+    train_mask = g.ndata["train_mask"]
+    optimizer = Adam(model.parameters(), args.lr, weight_decay=args.wd)
+
+    loss_fcn = nn.NLLLoss()
```
Comment (mufeili) [inline examples/sparse/gcnii.py:100]: For consistency with other examples, use `CrossEntropyLoss` with logits?

File: examples/sparse/gcnii.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        self.weight = Parameter(torch.FloatTensor(in_size, out_size))
+        self.reset_parameters()
+
+    ############################################################################
+    # (HIGHLIGHT) Take the advantage of DGL sparse APIs to implement the GCNII
+    # forward process.
+    ############################################################################
+    def forward(self, A_norm, H, H0, lamda, alpha, l):
+        beta = math.log(lamda / l + 1)
+        H = spmm(A_norm, H)
```
Comment (mufeili) [inline examples/sparse/gcnii.py:31]: Use `A_norm @ H` instead.

File: examples/sparse/gcnii.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+        self.CONV_layers = nn.ModuleList()
+        for _ in range(n_layers):
+            self.CONV_layers.append(GCNIIConvolution(hidden_size, hidden_size))
+        self.activation = nn.ReLU()
+        self.dropout = dropout
+
+    def forward(self, A_norm, feature):
+        H = feature
+        H = F.dropout(H, self.dropout)
```
Comment (mufeili) [inline examples/sparse/gcnii.py:70]:
```suggestion
        H = F.dropout(H, self.dropout, training = self.training)
```
to handle the different behavior during `train` and `eval`.

File: examples/sparse/gcnii.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        H = self.FC_layers[0](H)
+        H = self.activation(H)
+        H0 = H
+        for i, conv in enumerate(self.CONV_layers):
+            H = F.dropout(H, self.dropout)
```
Comment (mufeili) [inline examples/sparse/gcnii.py:75]: Ditto

**Review by mufeili — CHANGES_REQUESTED** [review]
Some minor comments

File: examples/sparse/gcnii.py (hunk 1 above)
Comment (paoxiaode) [reply examples/sparse/gcnii.py]: remove the argparse, add the config init in main function

File: examples/sparse/gcnii.py (hunk 2 above)
Comment (paoxiaode) [reply examples/sparse/gcnii.py]: use CrossEntropyLoss instead

File: examples/sparse/gcnii.py (hunk 4 above)
Comment (paoxiaode) [reply examples/sparse/gcnii.py]: add the  training = self.training

**Comment by frozenbugs** [comment]
@dgl-bot

File: examples/sparse/gcnii.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    ############################################################################
+    # (HIGHLIGHT) Take the advantage of DGL sparse APIs to implement the GCNII
+    # forward process.
+    ############################################################################
+    def forward(self, A_norm, H, H0, lamda, alpha, l):
+        beta = math.log(lamda / l + 1)
+
+        # Multiply a sparse matrix by a dense matrix.
+        H = A_norm @ H
+        support = (1 - alpha) * H + alpha * H0
```
Comment (frozenbugs) [inline examples/sparse/gcnii.py:31]: what's support?

File: examples/sparse/gcnii.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        alpha,
+        dropout=0.5,
+    ):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.n_layers = n_layers
+        self.lamda = lamda
+        self.alpha = alpha
+
+        self.FC_layers = nn.ModuleList()
```
Comment (frozenbugs) [inline examples/sparse/gcnii.py:53]: what does FC stand for?

File: examples/sparse/gcnii.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        self.FC_layers.append(nn.Linear(in_size, hidden_size))
+        self.FC_layers.append(nn.Linear(hidden_size, out_size))
+
+        self.CONV_layers = nn.ModuleList()
```
Comment (frozenbugs) [inline examples/sparse/gcnii.py:57]: Usually we suggest to use notation from the paper directly.

File: examples/sparse/gcnii.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        H = (1 - beta) * support + beta * self.weight(support)
+        return H
+
+
+class GCNII(nn.Module):
+    def __init__(
+        self,
+        in_size,
+        out_size,
+        hidden_size,
```
Comment (mufeili) [inline examples/sparse/gcnii.py:41]: Set default value for `hidden_size` here, etc.

**Comment by mufeili** [comment]
What accuracy number did you get?

File: examples/sparse/gcnii.py (hunk 7 above)
Comment (paoxiaode) [reply examples/sparse/gcnii.py:53]: in the paper, it first uses a FC layer to convert the init features into fixed size, then use GCNIIConv

File: examples/sparse/gcnii.py (hunk 6 above)
Comment (paoxiaode) [reply examples/sparse/gcnii.py:31]: In the paper, it is called Initial residual connection, I think mayber `res` is better?

**Comment by paoxiaode** [comment]
> What accuracy number did you get?

### Performace
|dataset |  val acc| test acc|
| - |   -|  -|
| Cora |  0.8 | 0.85 |

File: examples/sparse/gcnii.py (hunk 6 above)
Comment (paoxiaode) [reply examples/sparse/gcnii.py]: change support to H

**Review by mufeili — APPROVED** [review]
I'm good.

File: examples/sparse/gcnii.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        self.dropout = dropout
+
+    def forward(self, A_norm, feature):
+        H = feature
+        H = F.dropout(H, self.dropout, training=self.training)
+        H = self.layers[0](H)
+        H = self.activation(H)
+        H0 = H
+
+        # The CGNII convolution forward.
```
Comment (frozenbugs) [inline examples/sparse/gcnii.py:70]: GCNII

File: examples/sparse/gcnii.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        alpha,
+        dropout=0.5,
+    ):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.n_layers = n_layers
+        self.lamda = lamda
+        self.alpha = alpha
+
+        # The CGNII model.
```
Comment (frozenbugs) [inline examples/sparse/gcnii.py:53]: GCNII

File: examples/sparse/gcnii.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+def train(model, g, A_norm, H):
+    label = g.ndata["label"]
+    train_mask = g.ndata["train_mask"]
+    val_mask = g.ndata["val_mask"]
+    test_mask = g.ndata["test_mask"]
+    optimizer = Adam(model.parameters(), lr=0.01, weight_decay=5e-4)
+
+    loss_fcn = nn.CrossEntropyLoss()
+
+    for epoch in range(1500):
```
Comment (frozenbugs) [inline examples/sparse/gcnii.py:102]: 50 epoch should be enough?
Same to GCN, 200 is unnecessary. https://github.com/dmlc/dgl/blob/master/examples/sparse/gcn.py

File: examples/sparse/gcnii.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+        # Backward.
+        loss.backward()
+        optimizer.step()
+
+        # Evaluate the prediction.
+        val_acc, test_acc = evaluate(
+            model, A_norm, H, label, val_mask, test_mask
+        )
+        if epoch % 20 == 0:
```
Comment (frozenbugs) [inline examples/sparse/gcnii.py:120]: We can print every epoch for just 50 epoch.

**Review by frozenbugs — APPROVED** [review]

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
@dgl-bot
